drivers: modem: gsm: introduce lock

ycsin opened the pull request:
After the introduction of start/stop & rssi query work, these
additional functions can be invoked from different threads.

Things can quickly become hard to analyze when an interrupt
fires during one of these functions and upon return the
scheduler schedules another thread and run another function.

This PR introduces a simple mutex lock that guarantees thread
safety. Similar implementation can be found in hl7800 driver,
which has lots of public APIs that can be invoked from different
threads.

## Implementation

The idea is to use mutex to protect the work items which can be invoked by the gsm workq (thread 1), and necessary public APIs such as the `gsm_ppp_start` which can be invoked from thread 2 & the `gsm_ppp_stop` which can be invoked from thread 3

ycsin committed:
drivers: modem: gsm: move the rssi query an individual function

Move the rssi query part out from the rssi_handler work item as
an individual function, so that it can be used directly without
referencing the work.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
cc @bbilas @endian-benjamin

bbilas commented on drivers/modem/gsm_ppp.c:
```diff
@@ -571,11 +583,12 @@ static void set_ppp_carrier_on(struct gsm_modem *gsm)
 }
 
-static void rssi_handler(struct k_work *work)
+static void query_rssi(struct gsm_modem, *gsm, bool lock)
```
```suggestion
static void query_rssi(struct gsm_modem *gsm, bool lock)
```

ycsin replied:
typo, fixed, thanks

jukkar commented on drivers/modem/gsm_ppp.c:
```diff
@@ -571,10 +583,14 @@ static void set_ppp_carrier_on(struct gsm_modem *gsm)
 }
 
-static void query_rssi(struct gsm_modem *gsm)
+static void query_rssi(struct gsm_modem *gsm, bool lock)
```
Instead of a parameter to the function, imho it is more readable to have a locking version of the function.
So something like
```
static void query_rssi_lock(struct gsm_modem *gsm)
{
     gsm_ppp_lock(gsm);
     query_rssi(gsm);
     gsm_ppp_unlock(gsm);
}
```
This way user does not really need to remember what the extra bool parameter is for.

endian-benjamin commented:
Of course given this patch set, it is rather tempting to introduce a return value and a timeout param to `gsm_ppp_stop()` - or a new function possibly. I wouldn't insist on that change being done in this PR, but I'd prefer having `gsm_ppp_lock()` just take a timeout (you can just have all current calls pass `K_FOREVER`)

Another thing to note is that `gsm_ppp_stop()` is now blocking, which is sort of an implicit API breakage. Not sure if that's a big deal or not, just thought I'd point it out.

ycsin commented:
> introduce a return value and a timeout param to `gsm_ppp_stop()`

That is planned in the FSM implementation later

ycsin commented:
> Another thing to note is that `gsm_ppp_stop()` is now blocking, which is sort of an implicit API breakage. Not sure if that's a big deal or not, just thought I'd point it out.

The driver upstream is blocking since 19910666e03888e083244fe04246686d7531409f and more recently 60beb291dfbdde1e8aebf98dcb57a3dd8686cc4b

ycsin committed:
drivers: modem: gsm: introduce lock

After the introduction of start/stop & rssi query work, these
additional functions can be invoked from different threads.

Things can quickly become hard to analyze when an interrupt
fires during one of these functions and upon return the
scheduler schedules another thread and run another function.

This PR introduces a simple mutex lock that guarantees thread
safety. Similar implementation can be found in hl7800 driver,
which has lots of public APIs that can be invoked from different
threads.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

Update gsm_ppp.c

ycsin committed:
drivers: modem: gsm: remove mux_enabled flag and checks

mux_enabled check in the gsm_configure is unnecessary since it
is init and scheduled by gsm_ppp_start which means that the
mux must be disabled. The IS_ENABLED(CONFIG_GSM_MUX) check
should be good enough to determine whether or not the
mux_enable function should be ran.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

jukkar reviewed (approved):
LGTM
